src/dirwatcher.js

```diff
@@ -0,0 +1,107 @@
+'use strict';
+
+const fsPromises = require('fs/promises');
+const { EventEmitter } = require('events');
+
+class DirWatcher extends EventEmitter {
+  constructor() {
+    super();
+    this.directoryContents = {};
+    this.newDirectoryContents = {};
+    this.changedFiles = [];
+  }
+
+  async readDir(path) {
+    return fsPromises.readdir(path);
+  }
+
+  async readStats(path, fileNames) {
+    return fileNames.map(fileName => fsPromises.stat(`${path}/${fileName}`));
+  }
+
+  async saveState(path) {
+    // read dir file names
+    const fileNames = await this.readDir(path);
+    this.changedFiles = fileNames;
+    this.directoryContents > 0
+      ? this.emit('data change')
+      : this.emit('no data');
+    // read stats
+    const stats = await this.readStats(path, fileNames);
+    const statsResolved = await Promise.all(stats);
+
+    // combine stats
+    const statsArray = fileNames.map((el, index) => ({
+      [el]: statsResolved[index].mtime.toString().slice(0, 24),
+    }));
+    // return stats as state
+    return Object.assign({}, ...statsArray);
+  }
+
+  async watch(path, delay) {
+    this.directoryContents = await this.saveState(path);
+    // emit data change for first load if there is data:
+    if (Object.keys(this.directoryContents).length === 0) {
+      this.emit('no change');
+    } else {
+      this.emit('data change');
+    }
+
+    // loop
+    const run = async () => {
+      this.newDirectoryContents = await this.saveState(path);
+      const arrayOfNewFiles = Object.keys(this.newDirectoryContents);
+      const arrayOfOldFiles = Object.keys(this.directoryContents);
+
+      // 1. check for new files
+      const newFiles = arrayOfNewFiles
+        .filter(x => !arrayOfOldFiles.includes(x))
```

Comment:
Don't use `includes` if you already have `Map` or key-value paired object, since access to hashmap have `O(1)` complexity, but includes have `O(log(n))` complexity